File: BO/ssap-partner-portal-bo-frontend/src/types/ManHour.ts
export interface ISupporterOrganizationTotal {
  supporterOrganizationName: string
  manHour: number
}

export interface ISupporter {
  name: string
  manHour: number
}

export interface ISupporterOrganizationManHours {
  supporterOrganizationName: string
  supporters: ISupporter[]
}

export interface ISummaryManHourTypeHeader {
  supporterOrganizationManHours: ISupporterOrganizationManHours[]
  supporterOrganizationTotal: ISupporterOrganizationTotal[]
}

export interface IManHour {
  manHourTypeName: string
  subName: string
  serviceTypeName: string
  roleName: string
  supporterOrganizationTotal: ISupporterOrganizationTotal[]
  supporterOrganizationManHours: ISupporterOrganizationManHours[]
}

export interface IGetSummaryManHourTypeRequest {
  year: number
  month: number
}

export interface IGetSummaryManHourTypeResponse {
  yearMonth: string
  header: ISummaryManHourTypeHeader[]
  manHours: IManHour[]
}

export interface ISummaryManHour {
  direct: number
  pre: number
  sales: number
  ssap: number
  others: number
  total: number
}

export interface IContractTime {
  producer: number
  accelerator: number
  total: number
}
export interface ISummarySupporterManHour {
  supporterOrganizationId: string
  supporterOrganizationName: string
  supporterId: string
  supporterName: string
  summaryManHour: ISummaryManHour[]
  contractTime: IContractTime[]
  isConfirm: boolean
}
export interface IGetSummarySupporterManHoursRequest {
  year: number
  month: number
}

export interface IGetSummarySupporterManHoursResponse {
  yearMonth: string
  manHours: ISummarySupporterManHour[]
}

export interface IProject {
  projectId: string
  projectName: string
  customerId: string
  contractType: string
  thisMonthDirectSupportManHourMain: number
  thisMonthDirectSupportManHourSub: number
  thisMonthPreSupportManHour: number
  thisMonthContractTime: number
  totalProcessYPercent: number
}

export interface IGetSummaryServiceTypesManHoursResponse {
  serviceTypeId: string
  serviceTypeName: string
  directSupportManHourFactor: number
  projects: IProject[]
}

export interface IGetSummaryServiceTypesManHoursRequest {
  year: number
  month: number
}

export interface IGetSummarySupporterOrganizationsManHourResponse {
  supporterOrganizationId: string
  supporterOrganizationName: string
  annualSales: number
  monthlySales: number
  monthlyProjectPrice: number
  monthlyContractTime: number
  monthlyWorkTime: number
  monthlyWorkTimeRate: number
  monthlySupporters: number
  monthlyManHour: number
  monthlyOccupancyRate: number
  monthlyOccupancyTotalTime: number
  monthlyOccupancyTotalRate: number
  updateAt: string
}
export interface IGetSummarySupporterOrganizationsManHoursRequest {
  year: number
  month: number
  supporterOrganizationId?: string
}

export interface IMainSupporterUser {
  id: string
  name: string
}

export interface ISupporterUser {
  id: string
  name: string
}

export interface IAlertProject {
  projectId: string
  projectName: string
  customerId: string
  customerName: string
  supporterOrganizationId: string
  supporterOrganizationName: string
  serviceType: string
  serviceTypeName: string
  supportDateFrom: string
  supportDateTo: string
  totalContractTime: number
  thisMonthContractTime: number
  totalProfit: number
  thisMonthProfit: number
  mainSupporterUser: IMainSupporterUser
  supporterUsers: ISupporterUser[]
  summaryDirectSupportManHour: number
  summaryPreSupportManHour: number
  thisMonthDirectSupportManHour: number
  thisMonthPreSupportManHour: number
  summaryTheoreticalDirectSupportManHour: number
  summaryTheoreticalPreSupportManHour: number
  thisMonthTheoreticalDirectSupportManHour: number
  thisMonthTheoreticalPreSupportManHour: number
}

export interface IGetSummaryProjectManHourAlertsResponse {
  summaryThisMonthContractTime: number
  projects: IAlertProject[]
}

export interface IAlertManHour {
  supporterUserId: 'string'
  supporterUserName: 'string'
  role: 'string'
  thisMonthSupporterDirectSupportManHours: 0
  thisMonthSupporterPreSupportManHours: 0
  summarySupporterDirectSupportManHours: 0
  summarySupporterPreSupportManHours: 0
}

export interface IGetSummaryProjectManHourAlertRequest {
  year: number
  month: number
  serviceTypeId: string
}

export interface IGetSummaryProjectManHourAlertResponse {
  projectId: string
  projectName: string
  customerId: string
  customerName: string
  manHours: IAlertManHour[]
}

export interface ISupporterOrganization {
  id: string
  name: string
  shortName: string
}

export interface IGetSupporterOrganizationsRequest {
  year: number
  month: number
}
export interface IGetSupporterOrganizationsResponse {
  supporterOrganizations: ISupporterOrganization[]
}

export interface IServiceTypeItem {
  id: string
  name: string
}

export interface IGetServiceTypesRequest {
  year: number
  month: number
}

export interface IGetServiceTypesResponse {
  serviceTypes: IServiceTypeItem[]
}

export interface ISupporterUsers {
  id: string
  name: string
}

export interface ISummaryProjectManHourAlertsProjectItems {
  projectId: string
  projectName: string
  customerId: string
  customerName: string
  supporterOrganizationId: string
  supporterOrganizationName: string
  serviceType: string
  serviceTypeName: string
  supportDateFrom: string
  supportDateTo: string
  totalContractTime: string
  thisMonthContractTime: string
  totalProfit: string
  thisMonthProfit: string
  mainSupporterUser: IMainSupporterUser
  supporterUsers: ISupporterUsers[]
  summaryDirectSupportManHour: string
  summaryPreSupportManHour: string
  thisMonthDirectSupportManHour: number
  thisMonthPreSupportManHour: number
  summaryTheoreticalDirectSupportManHour: number
  summaryTheoreticalPreSupportManHour: number
  thisMonthTheoreticalDirectSupportManHour: number
  thisMonthTheoreticalPreSupportManHour: number
}

export interface IgetSummaryProjectManHourAlertsResponse {
  summaryThisMonthContractTime: number
  projects: ISummaryProjectManHourAlertsProjectItems[]
}

export interface IGetSummaryProjectManHourAlertsRequest {
  year: number
  month: number
  supporterOrganizationId?: string
  serviceTypeId?: string
}

export interface IManHours {
  supporterUserId: string
  supporterUserName: string
  role: string
  thisMonthSupporterDirectSupportManHours: number
  thisMonthSupporterPreSupportManHours: number
  summarySupporterDirectSupportManHours: number
  summarySupporterPreSupportManHours: number
}

export interface IGetSummaryProjectManHourAlertDetailResponse {
  projectId: string
  projectName: string
  customerId: string
  customerName: string
  manHours: IManHours[]
}

export interface IDirectSupportManHoursItem {
  projectId: string
  projectName: string
  role: string
  serviceType: string
  kartenManHour: number
  inputManHour: number
}
export interface IDirectSupportManHours {
  items: IDirectSupportManHoursItem[]
  memo: string
}

export interface IPreSupportManHoursItem {
  projectId: string
  projectName: string
  role: string
  serviceType: string
  inputManHour: number
}

export interface IPreSupportManHours {
  items: IPreSupportManHoursItem[]
  memo: string
}

export interface ISalesSupportManHoursItem {
  projectName: string
  customerId: string
  customerName: string
  type: string
  inputManHour: number
}

export interface ISalesSupportManHours {
  items: ISalesSupportManHoursItem[]
  memo: string
}

export interface ISsapManHours {
  meeting: number
  study: number
  learning: number
  newService: number
  startdash: number
  improvement: number
  ssap: number
  qc: number
  accounting: number
  management: number
  officeWork: number
  others: number
  memo: string
}

export interface IHolidaysManHours {
  paidHoliday: number
  holiday: number
  private: number
  others: number
  departmentOthers: number
  memo: string
}

export interface ISummarySupportManHours {
  direct: number
  pre: number
  sales: number
  ssap: number
  others: number
  total: number
}

export interface IGetManHourBySupporterUserIdRequest {
  year: number
  month: number
}

export interface IUpdateManHourRequestParams {
  id: string
  version: number
  year: number
  month: number
}
export interface IGetManHourBySupporterUserIdResponse {
  yearMonth: string
  supporterUserId: string
  supporterUserName: string
  supporterOrganizationId: string
  supporterOrganizationName: string
  directSupportManHours: IDirectSupportManHours[]
  preSupportManHours: IPreSupportManHours[]
  salesSupportManHours: ISalesSupportManHours[]
  ssapManHours: ISsapManHours
  holidaysManHours: IHolidaysManHours
  summarySupportManHours: ISummarySupportManHours
  isConfirm: boolean
  createId: string
  createUserName: string
  createAt: string
  updateId: string
  updateUserName: string
  updateAt: string
  version: number
}

export interface IUpdateManHourRequest {
  supporterUserName: string
  supporterOrganizationId: string
  supporterOrganizationName: string
  directSupportManHours: IDirectSupportManHours[]
  preSupportManHours: IPreSupportManHours[]
  salesSupportManHours: ISalesSupportManHours[]
  ssapManHours: ISsapManHours
  holidaysManHours: IHolidaysManHours
  summarySupportManHours: ISummarySupportManHours
  isConfirm: boolean
}

export interface IUpdateManHourResponse {
  yearMonth: string
  supporterUserId: string
  supporterUserName: string
  supporterOrganizationId: string
  supporterOrganizationName: string
  directSupportManHours: IDirectSupportManHours[]
  preSupportManHours: IPreSupportManHours[]
  salesSupportManHours: ISalesSupportManHours[]
  ssapManHours: ISsapManHours
  holidaysManHours: IHolidaysManHours
  summarySupportManHours: ISummarySupportManHours
  isConfirm: boolean
  createId: string
  createUserName: string
  createAt: string
  updateId: string
  updateUserName: string
  updateAt: string
  version: number
}

export interface IUpdateManHourBySupporterUserIdRequest {
  supporterUserName: string
  supporterOrganizationId: string
  supporterOrganizationName: string
  directSupportManHours: IDirectSupportManHours[]
  preSupportManHours: IPreSupportManHours[]
  salesSupportManHours: ISalesSupportManHours[]
  ssapManHours: ISsapManHours
  holidaysManHours: IHolidaysManHours
  isConfirm: boolean
}

export interface IFormattedMonthlyProjectOrganizationMember {
  isConfirm: true
  role: string
  supporterName: string
}

export interface IFormattedMonthlyProjectOrganization {
  supporterOrganizationId: string
  supporterOrganizationName: string
  members: IFormattedMonthlyProjectOrganizationMember[]
}

export interface IFormattedMonthlyProject {
  /**
   * supporterOrganizationId+'Total':number (支援者組織の合計人数)
   * supporterOrganizationId+'Producer':string (支援者組織のプロデューサ名),
   * supporterOrganizationId+'Accelerator':string (支援者組織のアクセラレータ名)
   * となるためやむなく [key: string]: unknown
   */
  [key: string]: unknown
  serviceTypeName: string
  name: string
  contractType: string
  mainSalesUserName: string
  organizations: IFormattedMonthlyProjectOrganization[]
  supportDateFrom: string
  supportDateTo: string
  totalContractTime: number
  thisMonthContractTime: string
  totalProfit: number
  thisMonthProfit: string
}

export const ENUM_FORMATTED_PROJECT = {
  INIT: 0,
  EXPECTATION: 1,
  EXCESS: 2,
} as const

File: pg/iret/FO/ssap-partner-portal-fo-backend/app/models/man_hour.py
from datetime import datetime

from app.core.config import get_app_settings
from app.models.custom_attribute.jst_datetime import JSTDateTimeAttribute
from pynamodb.attributes import (BooleanAttribute, ListAttribute, MapAttribute,
                                 NumberAttribute, UnicodeAttribute,
                                 VersionAttribute)
from pynamodb.indexes import AllProjection, GlobalSecondaryIndex
from pynamodb.models import Model


class YearMonthDataTypeIndex(GlobalSecondaryIndex):
    """GSI：年月データ区分検索"""

    class Meta:
        index_name = "year_month-data_type-index"
        projection = AllProjection()

    year_month = UnicodeAttribute(hash_key=True, null=False)
    data_type = UnicodeAttribute(range_key=True)


class SupportItemsSubAttribute(MapAttribute):
    """支援工数TBL 支援者別工数FASET  Attribute定義"""

    project_id = UnicodeAttribute(null=True)
    role = UnicodeAttribute(null=True)
    service_type = UnicodeAttribute(null=True)
    input_man_hour = NumberAttribute(null=True, default=0)


class AcquirementItemsSubAttribute(MapAttribute):
    """支援工数TBL 支援者別工数FASET  Attribute定義"""

    project_name = UnicodeAttribute(null=True)
    customer_name = UnicodeAttribute(null=True)
    type = UnicodeAttribute(null=True)
    input_man_hour = NumberAttribute(null=True, default=0)


class DirectSupportManHoursAttribute(MapAttribute):
    """支援工数TBL 支援者別工数FASET  Attribute定義"""

    items = ListAttribute(of=SupportItemsSubAttribute, null=True)
    memo = UnicodeAttribute(null=True)


class PreSupportManHoursAttribute(MapAttribute):
    """支援工数TBL 支援者別工数FASET  Attribute定義"""

    items = ListAttribute(of=SupportItemsSubAttribute, null=True)
    memo = UnicodeAttribute(null=True)


class SalesSupportManHoursAttribute(MapAttribute):
    """支援工数TBL 支援者別工数FASET  Attribute定義"""

    items = ListAttribute(of=AcquirementItemsSubAttribute, null=True)
    memo = UnicodeAttribute(null=True)


class SsapManHoursAttribute(MapAttribute):
    """支援工数TBL 支援者別工数FASET  Attribute定義"""

    meeting = NumberAttribute(null=True, default=0)
    study = NumberAttribute(null=True, default=0)
    learning = NumberAttribute(null=True, default=0)
    new_service = NumberAttribute(null=True, default=0)
    startdash = NumberAttribute(null=True, default=0)
    improvement = NumberAttribute(null=True, default=0)
    ssap = NumberAttribute(null=True, default=0)
    qc = NumberAttribute(null=True, default=0)
    accounting = NumberAttribute(null=True, default=0)
    management = NumberAttribute(null=True, default=0)
    office_work = NumberAttribute(null=True, default=0)
    others = NumberAttribute(null=True, default=0)
    memo = UnicodeAttribute(null=True)


class HolidaysManHoursAttribute(MapAttribute):
    """支援工数TBL 支援者別工数FASET  Attribute定義"""

    paid_holiday = NumberAttribute(null=True, default=0)
    holiday = NumberAttribute(null=True, default=0)
    private = NumberAttribute(null=True, default=0)
    others = NumberAttribute(null=True, default=0)
    department_others = NumberAttribute(null=True, default=0)
    memo = UnicodeAttribute(null=True)


class SummaryManHourAllowNullAttribute(MapAttribute):
    """支援工数TBL 支援者別工数FASET  Attribute定義"""

    direct = NumberAttribute(null=True, default=0)
    pre = NumberAttribute(null=True, default=0)
    sales = NumberAttribute(null=True, default=0)
    ssap = NumberAttribute(null=True, default=0)
    others = NumberAttribute(null=True, default=0)
    total = NumberAttribute(null=True, default=0)


class ManHourSupporterModel(Model):
    """支援工数TBL 支援者別工数FASET モデル定義"""

    class Meta:
        table_name = get_app_settings().table_prefix + "ManHour"
        region = get_app_settings().aws_region

    data_type = UnicodeAttribute(hash_key=True, null=False)
    year_month = UnicodeAttribute(range_key=True, null=False)
    supporter_user_id = UnicodeAttribute(null=False)
    supporter_name = UnicodeAttribute(null=False)
    supporter_organization_id = UnicodeAttribute(null=False)
    supporter_organization_name = UnicodeAttribute(null=False)
    direct_support_man_hours = DirectSupportManHoursAttribute(null=False)
    pre_support_man_hours = PreSupportManHoursAttribute(null=False)
    sales_support_man_hours = SalesSupportManHoursAttribute(null=False)
    ssap_man_hours = SsapManHoursAttribute(null=False)
    holidays_man_hours = HolidaysManHoursAttribute(null=False)
    summary_man_hour = SummaryManHourAllowNullAttribute(null=True)
    is_confirm = BooleanAttribute(null=False)

    create_id = UnicodeAttribute(null=True)
    create_at = JSTDateTimeAttribute(null=False, default=datetime.now())
    update_id = UnicodeAttribute(null=True)
    update_at = JSTDateTimeAttribute(null=False, default=datetime.now())
    version = VersionAttribute()

    year_month_data_type_index = YearMonthDataTypeIndex()


class ManHourSupporterOrganizationSummaryModel(Model):
    """支援工数TBL 支援者組織(課)別工数集計FASET モデル定義"""

    class Meta:
        table_name = get_app_settings().table_prefix + "ManHour"
        region = get_app_settings().aws_region

    data_type = UnicodeAttribute(hash_key=True, null=False)
    year_month = UnicodeAttribute(range_key=True, null=False)
    supporter_organization_id = UnicodeAttribute(null=False)
    supporter_organization_name = UnicodeAttribute(null=False)
    annual_sales = NumberAttribute(null=False, default=0)
    monthly_sales = NumberAttribute(null=False, default=0)
    monthly_project_price = NumberAttribute(null=False, default=0)
    monthly_contract_time = NumberAttribute(null=False, default=0)
    monthly_work_time = NumberAttribute(null=False, default=0)
    monthly_work_time_rate = NumberAttribute(null=False, default=0)
    monthly_supporters = NumberAttribute(null=False, default=0)
    monthly_man_hour = NumberAttribute(null=False, default=0)
    monthly_occupancy_rate = NumberAttribute(null=False, default=0)
    monthly_occupancy_total_time = NumberAttribute(null=False, default=0)
    monthly_occupancy_total_rate = NumberAttribute(null=False, default=0)

    create_at = JSTDateTimeAttribute(null=False, default=datetime.now())
    update_at = JSTDateTimeAttribute(null=False, default=datetime.now())

    year_month_data_type_index = YearMonthDataTypeIndex()


class SupporterUsersAttribute(MapAttribute):
    """支援工数TBL プロジェクト別工数集計FASET  Attribute定義"""

    id = UnicodeAttribute(null=True)
    name = UnicodeAttribute(null=True)
    is_confirm = BooleanAttribute(null=True)


class SupportManHourAttribute(MapAttribute):
    """支援工数TBL プロジェクト別工数集計FASET  Attribute定義"""

    supporter_user_id = UnicodeAttribute(null=False)
    input_man_hour = NumberAttribute(null=False, default=0)


class ManHourProjectSummaryModel(Model):
    """支援工数TBL プロジェクト別工数集計FASET モデル定義"""

    class Meta:
        table_name = get_app_settings().table_prefix + "ManHour"
        region = get_app_settings().aws_region

    data_type = UnicodeAttribute(hash_key=True, null=False)
    year_month = UnicodeAttribute(range_key=True, null=False)
    project_id = UnicodeAttribute(null=False)
    project_name = UnicodeAttribute(null=False)
    customer_id = UnicodeAttribute(null=False)
    customer_name = UnicodeAttribute(null=False)
    supporter_organization_id = UnicodeAttribute(null=False)
    service_type = UnicodeAttribute(null=False)
    support_date_from = UnicodeAttribute(null=True)
    support_date_to = UnicodeAttribute(null=True)
    contract_type = UnicodeAttribute(null=True)
    main_supporter_user = SupporterUsersAttribute(null=True)
    supporter_users = ListAttribute(of=SupporterUsersAttribute, null=True)
    total_contract_time = NumberAttribute(null=False)
    this_month_contract_time = NumberAttribute(null=True)
    total_profit = NumberAttribute(null=True)
    this_month_profit = NumberAttribute(null=True)
    this_month_direct_support_man_hour = NumberAttribute(null=True, default=0)
    this_month_direct_support_man_hour_main = NumberAttribute(null=True, default=0)
    this_month_direct_support_man_hour_sub = NumberAttribute(null=True, default=0)
    this_month_pre_support_man_hour = NumberAttribute(null=True, default=0)
    summary_direct_support_man_hour = NumberAttribute(null=True, default=0)
    summary_pre_support_man_hour = NumberAttribute(null=True, default=0)
    summary_theoretical_direct_support_man_hour = NumberAttribute(null=True, default=0)
    summary_theoretical_pre_support_man_hour = NumberAttribute(null=True, default=0)
    this_month_theoretical_direct_support_man_hour = NumberAttribute(
        null=True, default=0
    )
    this_month_theoretical_pre_support_man_hour = NumberAttribute(null=True, default=0)
    this_month_supporter_direct_support_man_hours = ListAttribute(
        of=SupportManHourAttribute, null=True
    )
    this_month_supporter_pre_support_man_hours = ListAttribute(
        of=SupportManHourAttribute, null=True
    )
    summary_supporter_direct_support_man_hours = ListAttribute(
        of=SupportManHourAttribute, null=True
    )
    summary_supporter_pre_support_man_hours = ListAttribute(
        of=SupportManHourAttribute, null=True
    )

    create_at = JSTDateTimeAttribute(null=False, default=datetime.now())
    update_at = JSTDateTimeAttribute(null=False, default=datetime.now())

    year_month_data_type_index = YearMonthDataTypeIndex()


class ManHourServiceTypeSummaryModel(Model):
    """支援工数TBL サービス種別別工数集計FASET モデル定義"""

    class Meta:
        table_name = get_app_settings().table_prefix + "ManHour"
        region = get_app_settings().aws_region

    data_type = UnicodeAttribute(hash_key=True, null=False)
    year_month = UnicodeAttribute(range_key=True, null=False)
    service_type_id = UnicodeAttribute(null=False)
    service_type_name = UnicodeAttribute(null=False)
    this_month_contract_time = NumberAttribute(null=False, default=0)
    this_month_paid_contract_time = NumberAttribute(null=False, default=0)
    number_of_supporters = NumberAttribute(null=True, default=0)
    man_hour_rate = NumberAttribute(null=True, default=0)
    total_man_hour = NumberAttribute(null=True, default=0)
    total_man_hour_y_percent = NumberAttribute(null=True, default=0)
    factor = NumberAttribute(null=True, default=0)

    create_at = JSTDateTimeAttribute(null=False, default=datetime.now())
    update_at = JSTDateTimeAttribute(null=False, default=datetime.now())

    year_month_data_type_index = YearMonthDataTypeIndex()
